### PR TITLE
Viewranger .VRC driver

### PR DESCRIPTION
**Pull request by andrew-aitchison**
## What does this PR do?
This is a driver which adds read-only support for the viewranger .VRC raster file format.
Raster maps in this format were formerly available from https://www.viewranger.com/en-gb
Although there is DRM and we attempt to preserve and make available all DRM information,
there is no known encrypted data in the file and no decryption in the driver.
.
## Tasklist

 - [ ] Write support for ViewRanger .VRC files
 - [ ] Should we also include support for ViewRanger .VRH height files?
 - [ ] Should we also include support for .VRV and .vmc files ? These are more about offline purchases.
 - [ ] Add test case(s)
 - [ ] Build on windows
 - [ ] Add documentation
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed
 - [ ] Review
 - [ ] Squash commits in a single one.

## Environment
Provide environment details, if relevant:

* OS:Ubuntu 21.04
* Compiler: GNU v7.3.1, GNU v11, Clang v12

### CODE REVIEW

File: gdal/frmts/vrc/VRC.cpp [hunk 1]
```diff
@@ -0,0 +1,4567 @@
...
+/************************************************************************/
+/*                              Identify()                              */
+/************************************************************************/
+
+int VRCDataset::Identify( GDALOpenInfo * poOpenInfo )
+
+{
+#if 1
+    const char* fileName = CPLGetFilename(poOpenInfo->pszFilename);
+    if( fileName==nullptr ) {
```
Comment (rouault) [inline gdal/frmts/vrc/VRC.cpp:1208]: that can't happen

File: gdal/frmts/vrc/VRC.cpp [hunk 2]
```diff
@@ -0,0 +1,4567 @@
...
+        return GDAL_IDENTIFY_FALSE;
+    }
+    if( !EQUAL(fileName + strlen(fileName) - strlen(".VRC"), ".VRC")) {
```
Comment (rouault) [inline gdal/frmts/vrc/VRC.cpp:1211]:
```suggestion
    if( !EQUAL(CPLGetExtension(fileName), "VRC")) {
```

File: gdal/frmts/vrc/VRC.cpp [hunk 3]
```diff
@@ -0,0 +1,4567 @@
...
+                 "Cannot allocate memory for tile index");
+        return nullptr;
+    }
+
+    // Read Tile Index into memory
+    // rotating it as we read it,
+    // since viewranger files start by going up the left column
+    // whilst gdal expects to go left to right across the top row.
+    for (int i=0; i<tileXcount; i++) {
+        int q = tileXcount*(tileYcount-1) +i;
```
Comment (rouault) [inline gdal/frmts/vrc/VRC.cpp:1296]: it should be checked that tileXcount * tileYcount fits on a int somewhere

File: gdal/frmts/vrc/VRC.cpp [hunk 4]
```diff
@@ -0,0 +1,4567 @@
...
+                     "VRCBuildTileIndex(0x%08x) tile %d 0x%08x: expected OverviewIndex with 7 entries - got %d",
+                     nTileIndexStart,
+                     nTileFound, nLastTileFound,
+                     nOverviewCount );
+            //VSIFree(anNewTileIndex);
+            //return nullptr;
+            break;
+        }
+        unsigned int anOverviewIndex[7]={};
+        size_t res=VSIFReadL(anOverviewIndex, 4, 7, fp);
```
Comment (rouault) [inline gdal/frmts/vrc/VRC.cpp:1426]:
```suggestion
        size_t res=VSIFReadL(anOverviewIndex, sizeof(unsigned int), 7, fp);
```

File: gdal/frmts/vrc/VRC.cpp [hunk 5]
```diff
@@ -0,0 +1,4567 @@
...
+
+{
+    CPLDebug("Viewranger",
+             "VRCDataset::Open( %p )",
+             poOpenInfo);
+    
+    if (!Identify(poOpenInfo))
+        return nullptr;
+
+#if defined(__GNUC__)
```
Comment (rouault) [inline gdal/frmts/vrc/VRC.cpp:1483]: all those debug statements should be removed

File: gdal/frmts/vrc/VRC.cpp [hunk 6]
```diff
@@ -0,0 +1,4567 @@
...
+                  " datasets.\n" );
+        return nullptr;
+    }
+
+
+/* -------------------------------------------------------------------- */
+/*      Create a corresponding GDALDataset.                             */
+/* -------------------------------------------------------------------- */
+
+    auto*poDS = new VRCDataset();
```
Comment (rouault) [inline gdal/frmts/vrc/VRC.cpp:1554]:
```suggestion
    auto poDS = std::unique_ptr<VRCDataset>(new VRTDataset());
```

File: gdal/frmts/vrc/VRC.cpp [hunk 7]
```diff
@@ -0,0 +1,4567 @@
...
+
+#if GDAL_VERSION_MAJOR >= 2
```
Comment (rouault) [inline gdal/frmts/vrc/VRC.cpp:1556]: to cleanup

File: gdal/frmts/vrc/VRC.cpp [hunk 8]
```diff
@@ -0,0 +1,4567 @@
...
+        GDALClose(poDS);
+        poDS=nullptr;
+        return nullptr;
+    }
+#endif
+
+/* -------------------------------------------------------------------- */
+/*      Read the header.                                                */
+/* -------------------------------------------------------------------- */
+    VSIFReadL( poDS->abyHeader, 1, 0x5a0, poDS->fp );
```
Comment (rouault) [inline gdal/frmts/vrc/VRC.cpp:1573]:
```suggestion
    VSIFReadL( poDS->abyHeader, 1, sizeof(poDS->abyHeader), poDS->fp );
```
?

File: gdal/frmts/vrc/VRC.cpp [hunk 9]
```diff
@@ -0,0 +1,4567 @@
...
+
+    poDS->nMagic = VRGetUInt(poOpenInfo->pabyHeader, 0);
+
+    if (poDS->nMagic != vrc_magic_metres && poDS->nMagic != vrc_magic_thirtysix) {
```
Comment (rouault) [inline gdal/frmts/vrc/VRC.cpp:1577]: check likely useless as already done in Identify()

File: gdal/frmts/vrc/VRC.cpp [hunk 10]
```diff
@@ -0,0 +1,4567 @@
...
+#else
+        CPLError(CE_Failure, CPLE_NotSupported,
+                 "VRC file %s unexpected Map ID %d",
+                 poOpenInfo->pszFilename, poDS->nMapID);
+        return nullptr;
+#endif
+    }
+
+    {
+#define VRCpszMapIDlen 11
```
Comment (rouault) [inline gdal/frmts/vrc/VRC.cpp:1647]:
```suggestion
        constexprt size_t VRCpszMapIDlen = 11;
```

File: gdal/frmts/vrc/VRC.cpp [hunk 11]
```diff
@@ -0,0 +1,4567 @@
...
+        char pszMapID[VRCpszMapIDlen]="";
+        int ret=CPLsnprintf(pszMapID, VRCpszMapIDlen, "0x%08x", poDS->nMapID);
+        pszMapID[VRCpszMapIDlen-1]='\000';
+        if(ret==VRCpszMapIDlen-1) {
+            poDS->SetMetadataItem("VRC ViewRanger MapID", pszMapID, "");
+        } else {
+            CPLDebug("Viewranger",
+                     "Could not set MapID Metadata - CPLsnprintf( , VRCpszMapIDlen, 0x%08x) returned %d", poDS->nMapID, ret);
+        }
+#undef VRCpszMapIDlen
```
Comment (rouault) [inline gdal/frmts/vrc/VRC.cpp:1657]:
```suggestion
```

File: gdal/frmts/vrc/VRC.cpp [hunk 12]
```diff
@@ -0,0 +1,4567 @@
...
+        if (*paszStrings[ii]) {
+            // Save the string as a MetadataItem.
+            const int VRCpszTAGlen=18;
+            char pszTag[VRCpszTAGlen+1]="";
+            int ret=CPLsnprintf(pszTag, VRCpszTAGlen, "String%u", ii);
+            pszTag[VRCpszTAGlen] = '\000';
+            if(VRCpszTAGlen>=ret && ret>0) {
+                poDS->SetMetadataItem
+                    ( pszTag,
+                      CPLRecode(paszStrings[ii],
```
Comment (rouault) [inline gdal/frmts/vrc/VRC.cpp:1706]: memory leak

File: gdal/frmts/vrc/VRC.cpp [hunk 13]
```diff
@@ -0,0 +1,4567 @@
...
+                      );
+            } else {
+                CPLDebug("Viewranger",
+                         "Could not set String%d Metadata - CPLsnprintf(..., VRCpszTAGlen %s) returned %d",
+                         ii, paszStrings[ii], ret);
+            }
+        }
+    }
+    if (nStringCount > 0) {
+        poDS->sLongTitle = CPLRecode(paszStrings[0],
```
Comment (rouault) [inline gdal/frmts/vrc/VRC.cpp:1717]: naming convention should be pszLongTitle. Similarly for sCopyright below

File: gdal/frmts/vrc/VRC.cpp [hunk 14]
```diff
@@ -0,0 +1,4567 @@
...
+        CPLError(CE_Failure, CPLE_AppDefined,
+                 "Map with %g metre pixels is too large scale (detailed) for the current VRC driver",
+                 poDS->dfPixelMetres);
+        return nullptr;        
+    }
+    
+    {
+        double dfRasterXSize =
+            ((10000.0)*(poDS->nRight - poDS->nLeft)) / poDS->nScale;
+        poDS->nRasterXSize = static_cast<int>(dfRasterXSize);
```
Comment (rouault) [inline gdal/frmts/vrc/VRC.cpp:1765]: this could potentially overflow. same below

**Review by rouault — COMMENTED** [review]
I've done a very partial review and stopped as it is overwhelming. Generally for upstream inclusion, you should remove lots of remaining #if for debugging. A lot of CPLDebug() are also probably no longer needed after initial developement stages. They add to the cognitive burden of reading the code.
If you need to iterate a lot on the current code to have it ready, you could potentially do that by closing this PR and then commit in your own fork where most CI tests (at least all github workflows) can be run, without "stealing" cycles for all remaining pull requests to OSGeo/gdal

**Comment by andrew-aitchison** [comment]
> If you need to iterate a lot on the current code to have it ready, you could potentially do that by closing this PR and 
> then commit in your own fork where most CI tests (at least all github workflows) can be run, without "stealing" 
> cycles for all remaining pull requests to OSGeo/gdal

Will do.

**Pull request closed without merging**